### PR TITLE
[WIP] Fast upgrade tikv pod

### LINKED ISSUE #3902: No need to wait tikv transfer leader timeout while upgrading tidb cluster with single tikv node
## Feature Request

**Is your feature request related to a problem? Please describe:**

It will take 10 minutes to wait tikv leader to transfer. Although there is only one tikv node. It's not necessary.

**Describe the feature you'd like:**

Upgrade the tikv directly if there is only one tikv

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by shonge**
### What problem does this PR solve?

Close #3902

### What is changed and how does it work?

When upgrade TiKV and stores number are less than 2, tidb-operator will update the tikv pod immediately.

### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [x] Unit test 
- [x] E2E test 
- [ ] Manual test 
- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
Fast upgrade TiKV when tikv replicas number are less than 2
```

### CODE REVIEW

**Comment by shonge** [comment]
@DanielZhangQD  PTAL

File: pkg/manager/member/tikv_upgrader.go [hunk 1]
```diff
@@ -209,6 +209,10 @@ func (u *tikvUpgrader) readyToUpgrade(upgradePod *corev1.Pod, tc *v1alpha1.TidbC
 			klog.Errorf("parse annotation:[%s] to time failed.", EvictLeaderBeginTime)
 			return false
 		}
+		if tc.Status.TiKV.StatefulSet.CurrentReplicas < 2 {
+			klog.Infof("TiKV stores are less than 2, skip waiting to evict region leader for Pod %s/%s", upgradePod.Namespace, upgradePod.Name)
+			return true
+		}
```
Comment (csuzhangxc) [inline pkg/manager/member/tikv_upgrader.go:215]: Can we move this block before L206?

**Comment by csuzhangxc** [comment]
BTW, if we have more TiKV pods but with only `1` replica in PD config, can we also fast upgrade TiKV too?

File: pkg/manager/member/tikv_upgrader.go (hunk 1 above)
Comment (DanielZhangQD) [reply pkg/manager/member/tikv_upgrader.go:215]: Do not use `CurrentReplicas` which is the number of Pods created by the StatefulSet controller from the StatefulSet version indicated by currentRevision. Here is an example during upgrade:
```
status:
  collisionCount: 0
  currentReplicas: 1
  currentRevision: t-tikv-5b4cd9d4df
  observedGeneration: 42
  readyReplicas: 2
  replicas: 3
  updateRevision: t-tikv-8444b66dc8
  updatedReplicas: 2
```

Comment (DanielZhangQD) [reply pkg/manager/member/tikv_upgrader.go:215]: BTW, should we move this block to the beginning of this function?

File: pkg/webhook/pod/tikv_deleter.go [hunk 2]
```diff
@@ -205,7 +205,7 @@ func (pc *PodAdmissionControl) admitDeleteUpTiKVPodDuringUpgrading(payload *admi
 		}
 	}
 
-	if !isTiKVReadyToUpgrade(payload.pod, store, tc.TiKVEvictLeaderTimeout()) {
+	if !isTiKVReadyToUpgrade(payload.pod, store, stores, tc.TiKVEvictLeaderTimeout()) {
```
Comment (DanielZhangQD) [inline pkg/webhook/pod/tikv_deleter.go:208]: What about checking the replicas as you're doing in the tikv_upgrader.go and only execute L193-L212 if replicas >= 2?

File: pkg/manager/member/tikv_upgrader.go (hunk 1 above)
Comment (DanielZhangQD) [reply pkg/manager/member/tikv_upgrader.go:215]: Refer to `tc.TiKVStsActualReplicas()`

**Comment by shonge** [comment]
> BTW, if we have more TiKV pods but with only `1` replica in PD config, can we also fast upgrade TiKV too?

I think we can fast upgrade TiKV when tc.spec.tikv.replicas=1 and tc.status.tikv.statefulSet.replicas more than 1.
We can transfer leaders to the smallest ordinal of tikv pods and waiting to evict leaders.

File: pkg/manager/member/tikv_upgrader.go (hunk 1 above)
Comment (shonge) [reply pkg/manager/member/tikv_upgrader.go]: update

File: pkg/webhook/pod/tikv_deleter.go (hunk 2 above)
Comment (shonge) [reply pkg/webhook/pod/tikv_deleter.go:214]: update

**Review by csuzhangxc — APPROVED** [review]
LGTM

File: pkg/manager/member/tikv_upgrader.go [hunk 3]
```diff
@@ -190,6 +190,11 @@ func (u *tikvUpgrader) upgradeTiKVPod(tc *v1alpha1.TidbCluster, ordinal int32, n
 func (u *tikvUpgrader) readyToUpgrade(upgradePod *corev1.Pod, tc *v1alpha1.TidbCluster) bool {
 	evictLeaderTimeout := tc.TiKVEvictLeaderTimeout()
 	tlsEnabled := tc.IsTLSClusterEnabled()
+	if tc.TiKVStsActualReplicas() < 2 {
```
Comment (DanielZhangQD) [inline pkg/manager/member/tikv_upgrader.go:193]: What about we checking the `len(podOrdinals)` in https://github.com/pingcap/tidb-operator/blob/86ea694ef0/pkg/manager/member/tikv_upgrader.go#L95 and just update the partition and return if the length < 2?

File: pkg/webhook/pod/tikv_deleter.go [hunk 4]
```diff
@@ -176,20 +176,26 @@ func (pc *PodAdmissionControl) admitDeleteUpTiKVPod(payload *admitPayload, store
 			klog.Infof("%s[%s/%s]'s tikv pod[%s/%s] failed to delete,%v", controllerKind, namespace, controllerName, namespace, name, err)
 			return util.ARFail(err)
 		}
-		return pc.admitDeleteUpTiKVPodDuringUpgrading(payload, store)
+		return pc.admitDeleteUpTiKVPodDuringUpgrading(payload, store, storesInfo)
```
Comment (DanielZhangQD) [inline pkg/webhook/pod/tikv_deleter.go:179]: Where the `storesInfo` is used now?

**Comment by DanielZhangQD** [comment]
Close via #4091.

**Pull request closed without merging**